File: atomic-x/ios/Classes/VideoRecorder/VideoRecorderHandler.swift
import Flutter
import UIKit
import SwiftUI

/**
 * VideoRecorderHandler
 * 处理 Flutter 层的 VideoRecorder 调用
 */
class VideoRecorderHandler: NSObject {
    private weak var viewController: UIViewController?
    private var pendingResult: FlutterResult?
    private let languageState = LanguageState()
    private let themeState = ThemeState()
    
    init(viewController: UIViewController?) {
        self.viewController = viewController
        super.init()
    }
    
    func handleStartRecord(call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("[VideoRecorderHandler] handleStartRecord called")
        
        guard pendingResult == nil else {
            print("[VideoRecorderHandler] VideoRecorder is already active")
            result(FlutterError(code: "ALREADY_ACTIVE",
                              message: "VideoRecorder is already active",
                              details: nil))
            return
        }
        
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS",
                              message: "Invalid arguments",
                              details: nil))
            return
        }
        
        pendingResult = result
        
        // 解析参数
        let recordModeInt = args["recordMode"] as? Int
        let videoQualityInt = args["videoQuality"] as? Int
        let minDurationMs = args["minDurationMs"] as? Int
        let maxDurationMs = args["maxDurationMs"] as? Int
        let isDefaultFrontCamera = args["isDefaultFrontCamera"] as? Bool
        let isSupportEdit = args["isSupportEdit"] as? Bool
        let isSupportBeauty = args["isSupportBeauty"] as? Bool
        let isSupportTorch = args["isSupportTorch"] as? Bool
        let isSupportAspect = args["isSupportAspect"] as? Bool
        let primaryColorHex = args["primaryColor"] as? String
        let languageCode = args["languageCode"] as? String
        let countryCode = args["countryCode"] as? String
        let scriptCode = args["scriptCode"] as? String
        
        print("[VideoRecorderHandler] Config - recordMode: \(recordModeInt ?? -1), videoQuality: \(videoQualityInt ?? -1), primaryColor: \(primaryColorHex ?? "nil"), language: \(languageCode ?? "nil")")
        
        // 设置语言
        if let languageCode = languageCode {
            setupLanguage(languageCode: languageCode, countryCode: countryCode, scriptCode: scriptCode)
        }
        
        // 设置主题色
        if let primaryColorHex = primaryColorHex, !primaryColorHex.isEmpty {
            print("[VideoRecorderHandler] Primary color: \(primaryColorHex)")
            themeState.setPrimaryColor(primaryColorHex)
        }
        
        // 创建配置
        let config = VideoRecorderConfig(
            recordMode: recordModeInt != nil ? RecordMode(rawValue: recordModeInt!) : nil,
            videoQuality: videoQualityInt != nil ? VideoQuality(rawValue: videoQualityInt! + 1) : nil,
            minDurationMs: minDurationMs,
            maxDurationMs: maxDurationMs,
            isDefaultFrontCamera: isDefaultFrontCamera,
            isSupportEdit: isSupportEdit,
            isSupportBeauty: isSupportBeauty,
            isSupportTorch: isSupportTorch,
            isSupportAspect: isSupportAspect
        )
        
        // 展示录制界面
        presentVideoRecorder(config: config)
    }
    
    private func setupLanguage(languageCode: String, countryCode: String?, scriptCode: String?) {
        // 规范化语言代码以匹配 .lproj 目录名
        var normalizedLanguage: String
        
        if languageCode.hasPrefix("zh") {
            // 中文需要 scriptCode 来区分简繁体
            if let scriptCode = scriptCode, !scriptCode.isEmpty {
                if scriptCode == "Hans" {
                    normalizedLanguage = "zh-Hans"  // 简体中文
                } else if scriptCode == "Hant" {
                    normalizedLanguage = "zh-Hant"  // 繁体中文
                } else {
                    normalizedLanguage = "zh-Hans"  // 默认简体
                }
            } else {
                normalizedLanguage = "zh-Hans"  // 默认简体
            }
        } else if languageCode.hasPrefix("en") {
            normalizedLanguage = "en"
        } else if languageCode.hasPrefix("ar") {
            normalizedLanguage = "ar"
        } else {
            normalizedLanguage = "en"  // 默认英文
        }
        
        print("[VideoRecorderHandler] Language set to: \(normalizedLanguage)")
        
        languageState.setLanguage(normalizedLanguage)
    }
    
    private func presentVideoRecorder(config: VideoRecorderConfig?) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let viewController = self.viewController else {
                self?.pendingResult?(FlutterError(code: "NO_VIEW_CONTROLLER",
                                                 message: "No view controller available",
                                                 details: nil))
                self?.pendingResult = nil
                return
            }
            
            let videoRecorderView = VideoRecorder(
                config: config,
                onVideoCaptured: { [weak self] filePath, durationMs, thumbnailPath in
                    self?.handleVideoCaptured(filePath: filePath, durationMs: durationMs, thumbnailPath: thumbnailPath)
                },
                onPhotoCaptured: { [weak self] filePath in
                    self?.handlePhotoCaptured(filePath: filePath)
                }
            )
            .environmentObject(self.themeState)
            
            let hostingController = UIHostingController(rootView: videoRecorderView)
            hostingController.modalPresentationStyle = .fullScreen
            
            viewController.present(hostingController, animated: true)
        }
    }
    
    private func handlePhotoCaptured(filePath: String?) {
        print("[VideoRecorderHandler] onPhotoCaptured: \(filePath ?? "nil")")
        
        // 销毁界面
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.viewController?.dismiss(animated: true)
        }
        
        let resultMap: [String: Any] = [
            "type": "photo",
            "filePath": filePath ?? ""
        ]
        
        pendingResult?(resultMap)
        pendingResult = nil
    }
    
    private func handleVideoCaptured(filePath: String?, durationMs: Int, thumbnailPath: String?) {
        print("[VideoRecorderHandler] onVideoCaptured: path=\(filePath ?? "nil"), duration=\(durationMs), thumbnail=\(thumbnailPath ?? "nil")")
        
        // 销毁界面
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.viewController?.dismiss(animated: true)
        }
        
        let resultMap: [String: Any] = [
            "type": "video",
            "filePath": filePath ?? "",
            "durationMs": durationMs,
            "thumbnailPath": thumbnailPath ?? ""
        ]
        
        pendingResult?(resultMap)
        pendingResult = nil
    }
}

File: atomic-x/android/src/main/kotlin/io/trtc/tuikit/atomicx/videorecorder/VideoRecorderHandler.kt
package io.trtc.tuikit.atomicx.videorecorder

import android.app.Application
import android.util.Log
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.trtc.tuikit.atomicx.basecomponent.theme.ThemeState
import io.trtc.tuikit.atomicx.basecomponent.utils.ContextProvider
import io.trtc.tuikit.atomicx.utils.LocaleUtils
import io.trtc.tuikit.atomicx.videorecorder.view.VideoRecorderActivity

class VideoRecorderHandler {
    companion object {
        private const val TAG = "VideoRecorderHandler"
    }

    private var pendingResult: MethodChannel.Result? = null
    private var lifecycleCallbacks: Application.ActivityLifecycleCallbacks? = null

    fun handleStartRecord(call: MethodCall, result: MethodChannel.Result) {
        Log.d(TAG, "handleStartRecord called")

        if (pendingResult != null) {
            Log.w(TAG, "VideoRecorder is already active")
            result.error("ALREADY_ACTIVE", "VideoRecorder is already active", null)
            return
        }

        try {
            pendingResult = result

            // 解析配置参数
            val recordModeInt = call.argument<Int>("recordMode")
            val videoQualityInt = call.argument<Int>("videoQuality")
            val minDurationMs = call.argument<Int>("minDurationMs")
            val maxDurationMs = call.argument<Int>("maxDurationMs")
            val isDefaultFrontCamera = call.argument<Boolean>("isDefaultFrontCamera")
            val isSupportEdit = call.argument<Boolean>("isSupportEdit")
            val isSupportBeauty = call.argument<Boolean>("isSupportBeauty")
            val isSupportRecordScrollFilter = call.argument<Boolean>("isSupportRecordScrollFilter")
            val isSupportTorch = call.argument<Boolean>("isSupportTorch")
            val isSupportAspect = call.argument<Boolean>("isSupportAspect")
            val primaryColor = call.argument<String>("primaryColor")
            val languageCode = call.argument<String>("languageCode")
            val countryCode = call.argument<String>("countryCode")
            val scriptCode = call.argument<String>("scriptCode")

            Log.d(TAG, "Config - recordMode: $recordModeInt, videoQuality: $videoQualityInt, primaryColor: $primaryColor, language: $languageCode")

            // 设置主题色
            if (!primaryColor.isNullOrEmpty()) {
                ThemeState.shared.setPrimaryColor(primaryColor)
            }

            // 设置语言
            if (!languageCode.isNullOrEmpty()) {
                setupLanguageCallback(languageCode, countryCode, scriptCode)
            }

            // 构建配置
            val config = VideoRecorderConfig(
                recordMode = recordModeInt?.let { RecordMode.fromInteger(it) },
                videoQuality = videoQualityInt?.let { 
                    // Flutter 枚举的 index 从 0 开始，但 VideoQuality 的 value 从 1 开始
                    VideoQuality.fromInteger(it + 1) 
                },
                minDurationMs = minDurationMs,
                maxDurationMs = maxDurationMs,
                isDefaultFrontCamera = isDefaultFrontCamera,
                isSupportEdit = isSupportEdit,
                isSupportBeauty = isSupportBeauty,
                isSupportRecordScrollFilter = isSupportRecordScrollFilter,
                isSupportTorch = isSupportTorch,
                isSupportAspect = isSupportAspect
            )

            // 调用原生录制
            VideoRecorder.startRecord(config, object : VideoRecordListener {
                override fun onPhotoCaptured(filePath: String?) {
                    Log.d(TAG, "onPhotoCaptured: $filePath")
                    
                    val resultMap = mapOf(
                        "type" to "photo",
                        "filePath" to (filePath ?: "")
                    )
                    
                    pendingResult?.success(resultMap)
                    pendingResult = null
                    cleanupLanguageCallback()
                }

                override fun onVideoCaptured(filePath: String?, durationMs: Int, thumbnailPath: String?) {
                    Log.d(TAG, "onVideoCaptured: path=$filePath, duration=$durationMs, thumbnail=$thumbnailPath")
                    
                    val resultMap = mapOf(
                        "type" to "video",
                        "filePath" to (filePath ?: ""),
                        "durationMs" to durationMs,
                        "thumbnailPath" to (thumbnailPath ?: "")
                    )
                    
                    pendingResult?.success(resultMap)
                    pendingResult = null
                    cleanupLanguageCallback()
                }
            })

        } catch (e: Exception) {
            Log.e(TAG, "Error in handleStartRecord", e)
            pendingResult?.error("VIDEO_RECORDER_ERROR", e.message, null)
            pendingResult = null
            cleanupLanguageCallback()
        }
    }

    private fun setupLanguageCallback(languageCode: String, countryCode: String?, scriptCode: String?) {
        val application = ContextProvider.appContext as? Application ?: return
        
        lifecycleCallbacks = LocaleUtils.registerLanguageCallback(
            application = application,
            targetActivityClass = VideoRecorderActivity::class.java,
            languageCode = languageCode,
            countryCode = countryCode,
            scriptCode = scriptCode,
            onActivityDestroyed = {
                cleanupLanguageCallback()
            }
        )
    }

    private fun cleanupLanguageCallback() {
        val application = ContextProvider.appContext as? Application
        application?.let {
            LocaleUtils.unregisterLanguageCallback(it, lifecycleCallbacks)
            lifecycleCallbacks = null
        }
    }
}

